### PR TITLE
Enable MapUserProvider storing username with the letter case signific…

### LINKED ISSUE #10245: Enable MapUserProvider storing username with the letter case significance
https://github.com/keycloak/keycloak/discussions/10198

> The user storage map provider should specify an option whether username letter case is significant or not.

> It needs to be stressed in Javadoc and later the documentation that the lettercase significance depends on storage.

> This option should be introduced on the level of [map storage user provider factory](https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java) and applied in appropriate [criteria builder](https://github.com/keycloak/keycloak/blob/90d4e586b6ce92c0b4211586ad554a5a07566417/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java#L56), not [map users provider factory](https://github.com/keycloak/keycloak/discussions/model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java), so that it is possible to configure the behaviour per realm.

### PR DESCRIPTION
**Pull request by vramik**
…ance

Closes #10245

### CODE REVIEW

**Review by sguilhen — COMMENTED** [review]
@vramik Thanks for getting into this. I've reviewed the whole thing and I must say I find it a bit confusing for the user provider to check a property set in the realm storage config. I say this because the storages themselves (JPA, hotrod, etc) never actually use this property and it feels like this switch has been put in the wrong place.

Perhaps I'm failing to see how this is supposed to work with realm-specific stores, but I have to say that this switch is really confusing to me.

**Comment by vramik** [comment]
> @vramik Thanks for getting into this. I've reviewed the whole thing and I must say I find it a bit confusing for the user provider to check a property set in the realm storage config. I say this because the storages themselves (JPA, hotrod, etc) never actually use this property and it feels like this switch has been put in the wrong place.

You're right it was misplaced at that stage of the draft PR. The setting now is taken from realm attribute (if the realm is available) upon creation of user provider. Default setting is `case-insensitive`.

The implementation follows https://github.com/keycloak/keycloak/discussions/10198#discussioncomment-3228656 and also incorporates https://github.com/keycloak/keycloak/discussions/10198#discussioncomment-3230123

@sguilhen @mhajas @hmlnarik could you please take a look?

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @vramik, I added some comments

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java [hunk 1]
```diff
@@ -44,6 +44,8 @@ abstract class AbstractUserEntity extends Impl implements MapUserEntity {
 
         private static final Logger LOG = Logger.getLogger(MapUserProvider.class);
         private String id;
+        private String username;
+        private String usernameLowerCase;
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java:48]: > If I understand correctly, is this field only used when serializing to JSON? 

No, actually the field is used only when querying by username (in case of `case-insensitive` search).

> Not sure how difficult is to compute lowercase version of String

Not sure exactly myself atm, it'd have to be updated in classes which actually parses the changesets to allow include a function as well (if it is not allowed already)

> but if we add a getter we could use it also in `MapFieldPredicates` as I suggested in the commit I sent you yesterday.

I'm not sure I follow here. Do you mean:

```
put(USER_PREDICATES, UserModel.SearchableFields.USERNAME, u -> (u instanceof MapUserEntityImpl) ? ((MapUserEntityImpl) u).getUsernameLowerCase() : KeycloakModelUtils.toLowerCaseSafe(u.getUsername()));
```
? If so I'd rather keep current state as it seems to me much cleaner. Maybe I am missing something here.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java [hunk 2]
```diff
@@ -40,12 +40,24 @@
...
     public MapUserProvider createNew(KeycloakSession session) {
+        RealmModel realm = session.getContext().getRealm();
+        boolean usernameCaseSensitive = false; //default to false
+
+        // if a realm is available take it from there
+        if (realm != null) {
+            usernameCaseSensitive = realm.getAttribute(USERNAME_CASE_SENSITIVE, Boolean.FALSE);
+        }
+
+        session.setAttribute(USERNAME_CASE_SENSITIVE, usernameCaseSensitive);
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java:59]: > Why do we configure case-sensitiveness per realm? 

The requirement comes from the discussion. See https://github.com/keycloak/keycloak/discussions/10198#discussioncomment-2179669 for the reference.

> What if someone disables case-sensitiveness when storage contains conflicting usernames? Should we have this switch somehow more static/fixed as part of the storage config maybe?

I asked similar question here: https://github.com/keycloak/keycloak/pull/10332#issuecomment-1169172848 

> Why are we storing this value in `session`? If case-sensitiveness is configured per realm, what if we do user search per more realms within one `session`? For example, search admin user in realm1 and then realm2? Can we just do `realm.getAttribute` in `MapUserProvider`?

Good point. We've discussed it with @hmlnarik in one of the sync calls and agreed that the session will be used. I don't recall exactly the reason right now, but it's true it won't work as it now with more realms within one session. I was able to reproduce it in model test. Right now I'm thinking whether it can happen also in 'real' scenario, when you use rest endpoints. Anyway I'm happy to update it. @hmlnarik could you please comment on this? Thank you

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java:48]: > > If I understand correctly, is this field only used when serializing to JSON?
> 
> No, actually the field is used only when querying by username (in case of `case-insensitive` search).

Hmm, and when is the lowercase version of the username stored in the database? I thought the `usernameLowercase` field is somehow projected to `usernameLowercase` within the database and that is why we need it.

> > but if we add a getter we could use it also in `MapFieldPredicates` as I suggested in the commit I sent you yesterday.
> 
> I'm not sure I follow here. Do you mean:
> 
> ```
> put(USER_PREDICATES, UserModel.SearchableFields.USERNAME, u -> (u instanceof MapUserEntityImpl) ? ((MapUserEntityImpl) u).getUsernameLowerCase() : KeycloakModelUtils.toLowerCaseSafe(u.getUsername()));
> ```
> 
> ? If so I'd rather keep current state as it seems to me much cleaner. Maybe I am missing something here.

Well, in this case, the lowercase version of the username is computed each time the searchable field is used and not only in the CHM but also in HotRod as MapFieldPredicates are partially used also there. But I am not sure how big the problem is.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java:59]: Thanks for the explanation @vramik, sorry I missed the discussion points. Let's wait for Hynek for feedback.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java:48]: > Hmm, and when is the lowercase version of the username stored in the database? I thought the `usernameLowercase` field is somehow projected to `usernameLowercase` within the database and that is why we need it.

Sorry, maybe I didn't express myself good enough. As the field is part of the `MapUserEntity` and the `JpaUserMetadata` extends `MapUserEntityImpl` it is already part of the json. Then it's generated into separate column in the database. And the column is then used in the search. Hope it makes more sense.

> Well, in this case, the lowercase version of the username is computed each time the searchable field is used and not only in the CHM but also in HotRod as MapFieldPredicates are partially used also there. But I am not sure how big the problem is.

If it'd be a problem I'd prefer to use the approach with `IgnoreMissingSetterForEntityClonerGenerator` annotation and use 
```
put(USER_PREDICATES, UserModel.SearchableFields.USERNAME, MapUserEntity::getUsernameLowerCase);
```

**Review by sguilhen — COMMENTED** [review]
I know this is still pending some feedback from @hmlnarik , but I have to say that overall this looks a lot better and I like it that the config is now coming from a realm attribute.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java (hunk 1 above)
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java:48]: Each implementation should be able to decide which way it would implement lettercase (in)sensitive search. For that reason, there should be no lowercase username variant prescribed in the generic map entity (nor in the AbstractUserEntity).

If that is needed say in JPA, then JPA entity metadata object should extend the `AbstractUserEntity` and add lowercase variant there. Yet - can we do without it?

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you @vramik, this PR is going in the right direction.

Please see a few inline comments. After addressing the two main comments re the session attribute and lowercase variant of the username, it should be good to be merged

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java]: Unfortunately there is not always appropriate realm in the context. I was able to set the correct realm at some places, but then I discovered a scenario where it wouldn't be easy. E.g. during importing a realm there is 

1. realm admin request authenticated for "master" realm where identity token is being verified for "admin" user to check whether the user has permission to create new realms
3. it leads to a creation of new instance of `UserProvider` (`MapUserProvider`) within the session, where case-sensitivity attribute for "master" realm would be added into the session attributes
4. Then when request lands into `RealmsAdminResource.importRealm(RealmRepresentation)` trying to create new realm, there is already provider created in the session, so no new provider is being created where it could be the attribute for the new realm inserted into session
5. when new users are tried to be created it's checked whether the user exist already and this call fail due to the above

One possible solution could be not to create new provider only within the session but also for each realm. 

As discussed with @mhajas ... for now I'll move the logic into `MapUserProvider` only and check directly the realm attribute only when it's needed for the particular search.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java (hunk 1 above)
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserEntity.java]: > Yet - can we do without it?

It's not possible to create generated column from already generated column, therefore I've added the the field into `JpaUserEntity`. Or did you have something else in mind?

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java]: Thanks for the investigation @vramik. Personally, I like the solution this way. It would make sense to have it in the session if the realm is always the same within each request, but since it is not the case, as you found out, I lean towards the solution you proposed in the latest changes.

If we want to optimize the obtaining of the lettercase flag from the realm, maybe we should take a step back and think about deploying some generic solution for all realm attributes. WDYT? 

Some ideas already here: https://github.com/keycloak/keycloak/discussions/12328

[the remark this answers is not in the document]
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java]: > If we want to optimize the obtaining of the lettercase flag from the realm, maybe we should take a step back and think about deploying some generic solution for all realm attributes. WDYT?
> 
> Some ideas already here: #12328

Thanks, agreed.

File: model/map-jpa/src/main/resources/META-INF/users/jpa-users-changelog-1.xml [hunk 3]
```diff
@@ -32,6 +32,9 @@ limitations under the License.
                 <constraints nullable="false"/>
             </column>
             <column name="metadata" type="json"/>
+            <column name="usernamelowercase" type="VARCHAR(255)">
+                <constraints nullable="false"/>
+            </column>
```
Comment (hmlnarik) [inline model/map-jpa/src/main/resources/META-INF/users/jpa-users-changelog-1.xml:37]: Could changes in this file be placed in a separate changeset?

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java [hunk 4]
```diff
@@ -40,6 +40,8 @@
  */
 public class MapUserProviderFactory extends AbstractMapProviderFactory<MapUserProvider, MapUserEntity, UserModel> implements UserProviderFactory<MapUserProvider>, InvalidationHandler {
 
+    public static final String USERNAME_CASE_SENSITIVE = "username-case-sensitive";
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserProviderFactory.java:43]: The attribute name should be clearly keycloak internal, e.g.:
```suggestion
    public static final String REALM_ATTR_USERNAME_CASE_SENSITIVE = "keycloak.username-search.case-sensitive";
```

File: server-spi/src/main/java/org/keycloak/models/UserModel.java [hunk 5]
```diff
@@ -54,6 +54,7 @@ public static class SearchableFields {
         public static final SearchableModelField<UserModel> ID              = new SearchableModelField<>("id", String.class);
         public static final SearchableModelField<UserModel> REALM_ID        = new SearchableModelField<>("realmId", String.class);
         public static final SearchableModelField<UserModel> USERNAME        = new SearchableModelField<>("username", String.class);
+        public static final SearchableModelField<UserModel> USERNAME_CASE_INSENSITIVE = new SearchableModelField<>("usernameCaseInsensitive", String.class);
```
Comment (hmlnarik) [inline server-spi/src/main/java/org/keycloak/models/UserModel.java:57]: These two searchable fields (`USERNAME`, `USERNAME_CASE_INSENSITIVE`) deserve javadoc which states which one is used for case (in)sensitive search

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 6]
```diff
@@ -93,6 +94,10 @@ public MapUserProvider(KeycloakSession session, MapStorage<MapUserEntity, UserMo
         session.getTransactionManager().enlist(tx);
     }
 
+    private Boolean getUsernameCaseSensitiveAttribute(RealmModel realm) {
+        return realm.getAttribute(USERNAME_CASE_SENSITIVE, Boolean.FALSE);
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:98]: The default value (`Boolean.FALSE`) should also be a constant in `MapUserProviderFactory`.

**Review by hmlnarik — COMMENTED** [review]
Would it be possible to add integration tests which scenarios where
- realm with case sensitivity search is (a) enabled, (b) disabled
- create a user
- search for the user by (a) uppercase, (b) original letter case, (c) lowercase username
- switch the case sensitivity flag to the other state
- search as previously

File: model/map-jpa/src/main/resources/META-INF/users/jpa-users-changelog-1.xml (hunk 3 above)
Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/users/jpa-users-changelog-1.xml:37]: Definitely, what would be a benefit to have it separated?

File: server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java [hunk 7]
```diff
@@ -271,10 +271,6 @@ private Map<String, List<String>> normalizeAttributes(Map<String, ?> attributes)
                     values = (List<String>) value;
                 }
 
-                if (key.equals(UserModel.USERNAME)) {
-                    values = Collections.singletonList(values.get(0).toLowerCase());
-                }
-
```
[the remark this answers is not in the document]
Comment (vramik) [reply server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java:277]: Without this change new user would be created with its username in lowered case format.

File: model/map-jpa/src/main/resources/META-INF/users/jpa-users-changelog-1.xml (hunk 3 above)
Comment (hmlnarik) [reply model/map-jpa/src/main/resources/META-INF/users/jpa-users-changelog-1.xml:37]: Those already trying the JPA storage would not be confused by liquibase conflicts.

Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/users/jpa-users-changelog-1.xml:37]: True, on the other hand there is opened https://github.com/keycloak/keycloak/pull/13392 which renames column names across whole jpa map storage and it will affect all checksums as well. So we may not be able to avoid the confusion. wdyt?

File: server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmlnarik) [reply server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java:277]: I can now see that this is redundant remnant that should be only implemented in the storage layer and as such, this is fixing a bug. Thanks

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedActionTotpSetupTest.java [hunk 8]
```diff
@@ -116,12 +117,13 @@ public void setupTotpRegister() {
...
         events.poll(); // skip to totp event
-        String authSessionId = events.expectRequiredAction(EventType.UPDATE_TOTP).user(userId).detail(Details.USERNAME, "setuptotp").assertEvent()
+        String expectedString = keycloakUsingProviderWithId(UserProvider.class, "jpa") ? "setuptotp" : "setupTotp";
+        String authSessionId = events.expectRequiredAction(EventType.UPDATE_TOTP).user(userId).detail(Details.USERNAME, expectedString).assertEvent()
                 .getDetails().get(Details.CODE_ID);
 
         assertKcActionStatus(SUCCESS);
         
-        events.expectLogin().user(userId).session(authSessionId).detail(Details.USERNAME, "setuptotp").assertEvent();
+        events.expectLogin().user(userId).session(authSessionId).detail(Details.USERNAME, expectedString).assertEvent();
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedActionTotpSetupTest.java:126]: Could it be the following instead?
```suggestion
        events.expectLogin().user(userId).session(authSessionId).detail(Details.USERNAME, expectedString).usernameIgnoringLettercase().assertEvent();
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/AccountRestServiceTest.java [hunk 9]
```diff
@@ -393,7 +394,8 @@ public void testUpdateProfile() throws IOException {
 
             user.setUsername("updatedUsername");
             user = updateAndGet(user);
-            assertEquals("updatedusername", user.getUsername());
+            String expectedString = keycloakUsingProviderWithId(UserProvider.class, "jpa") ? "updatedusername" : "updatedUsername";
+            assertEquals(expectedString, user.getUsername());
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/AccountRestServiceTest.java:398]: Could it be
```suggestion
            assertThat("updatedusername", Matchers.equalToIgnoringCase(user.getUsername()));
```
?

**Review by mhajas — DISMISSED** [review]
Thanks @vramik! Great work. I like it this way. There is one thing that caught my attention, but it is mostly my curiosity because we are doing quite the opposite in HotRod.

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java [hunk 10]
```diff
@@ -306,6 +306,33 @@ private static MapModelCriteriaBuilder<Object, MapGroupEntity, GroupModel> check
...
+
+        final Object ob = values[0];
+        if (!String.class.isAssignableFrom(ob.getClass())) {
+            throw new CriterionNotSupportedException(UserModel.SearchableFields.USERNAME_CASE_INSENSITIVE, op, "Invalid arguments, expected (String), got: " + Arrays.toString(values));
+        }
+
+        String value = (String) ob;
+        Function<MapUserEntity, ?> getter;
+
+        switch (op) {
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java:322]: It's implemented this way in `Jpa*ModelCriteriaBuilder`s. There it saves time when some operator is not implemented and is tried to be used then it fails fast with clear message. On the other hand if there is a way how to implement it for more generically it may be worth changing in a way you're proposing.

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 11]
```diff
@@ -488,42 +496,51 @@ public UserModel getUserByUsername(RealmModel realm, String username) {
...
+                LOG.warnf("There are colliding usernames for users with usernames and ids: %s", 
+                        users.stream().collect(Collectors.toMap(MapUserEntity::getUsername, MapUserEntity::getId)));
+            }
+            return entityToAdapterFunc(realm).apply(users.get(0));
         }
     }
 
     @Override
     public UserModel getUserByEmail(RealmModel realm, String email) {
+        //emails are stored in lowered case format in map storage, see MapUserAdapter.setEmail(String email)
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:517]: Does this apply to external storages like LDAP?

**Comment by hmlnarik** [comment]
undertow-map-jpa, group1 failed due to #14270

**Review by hmlnarik — APPROVED** [review]
Thanks for the PR!
